src/stdlib/replace_with.rs

```diff
@@ -0,0 +1,214 @@
+use std::collections::BTreeMap;
+
+use crate::compiler::prelude::*;
+
+fn replace_with<T>(
+    value: Value,
+    pattern: Value,
+    count: Value,
+    ctx: &mut Context,
+    runner: closure::Runner<T>,
+) -> Resolved
+where
+    T: Fn(&mut Context) -> Result<Value, ExpressionError>,
+{
+    let haystack = value.try_bytes_utf8_lossy()?;
+    let count = match count.try_integer()? {
+        i if i > 0 => i as usize,
+        i if i < 0 => 0,
+        // this is when i == 0
+        _ => return Ok(value),
+    };
+    match pattern {
+        Value::Regex(regex) => {
+            let captures = regex.captures_iter(&haystack);
+            make_replacement(captures, &haystack, count, ctx, runner)
+        }
+        value => Err(ValueError::Expected {
+            got: value.kind(),
+            expected: Kind::regex(),
+        }
+        .into()),
+    }
+}
+
+fn make_replacement<T>(
+    caps: regex::CaptureMatches,
+    haystack: &str,
+    count: usize,
+    ctx: &mut Context,
+    runner: closure::Runner<T>,
+) -> Resolved
+where
+    T: Fn(&mut Context) -> Result<Value, ExpressionError>,
+{
+    // possible optimization: peek at first capture, if none return the original value.
+    let mut replaced = String::with_capacity(haystack.len());
+    let limit = if count == 0 { usize::MAX } else { count - 1 };
+    let mut last_match = 0;
+    // we loop over the matches ourselves instead of calling Regex::replacen, so that we can
+    // handle errors. This is however based on the implementation of Regex::replacen
+    for (i, cap) in caps.enumerate() {
+        // Safe to unrap because the 0th index always includes the full match.
+        let m = cap.get(0).unwrap(); // full match
+
+        let mut value = captures_to_value(&cap);
+        runner.map_value(ctx, &mut value)?;
+        let replacement = value.try_bytes_utf8_lossy()?;
+
+        replaced.push_str(&haystack[last_match..m.start()]);
+        replaced.push_str(&replacement);
+        last_match = m.end();
+        if i >= limit {
+            break;
+        }
+    }
+    // add the final component
+    replaced.push_str(&haystack[last_match..]);
+    Ok(replaced.into())
+}
+
+fn captures_to_value(captures: &regex::Captures) -> Value {
```

Comment:
Is using an array the best representation for the match? would it be better to use an object with  a key for the whole match, a key fro the capture groups, and maybe something for the named capture groups?

Reply:
Perhaps an object would be cleaner. Up to you.

Reply:
I decided to go with an object, so that we can expose the named capture groups by name. 

The downsides of this approach are:
- If there are named capture groups, we have to clone the strings (I wonder if an `Arc<str>` instead of a string for a string value would benefit performance)
- It's a little more verbose to reference non-named capture groups